test/affinity_tests.cc

```diff
@@ -0,0 +1,323 @@
+		CHECK(cfg.enabled == enabled);
+		CHECK(cfg.starting_from_core == starting_from_core);
+		CHECK(cfg.hardcoded_core_ids == hardcoded_core_ids);
+	};
+
+	SECTION("when unset") {
+		// we expect unset to be auto
+		const auto cfg = detail::thread_pinning::parse_validate_env(""sv);
+		check_cfg(cfg, true, 1, {});
+	}
+
+	SECTION("when auto") {
+		const auto cfg = detail::thread_pinning::parse_validate_env("auto"sv);
+		check_cfg(cfg, true, 1, {});
+	}
+
+	SECTION("when from:3") {
+		const auto cfg = detail::thread_pinning::parse_validate_env("from:3"sv);
+		check_cfg(cfg, true, 3, {});
+	}
+
+	SECTION("when 1,2,3") {
+		const auto cfg = detail::thread_pinning::parse_validate_env("1,2,3"sv);
+		check_cfg(cfg, true, 0, {1, 2, 3});
+	}
+
+	SECTION("when bool") {
+		const auto cfg = detail::thread_pinning::parse_validate_env("false"sv);
+		check_cfg(cfg, false, 1, {});
+	}
+}
+
+TEST_CASE("thread pinning environment error handling", "[affinity][config]") {
+	test_utils::allow_max_log_level(detail::log_level::warn);
+	using namespace std::string_view_literals;
+
+	const auto error_string = "Cannot parse CELERITY_THREAD_PINNING setting";
+
+	SECTION("for the 'from' case") { CHECK_THROWS_WITH(detail::thread_pinning::parse_validate_env("from:a"sv), Catch::Matchers::StartsWith(error_string)); }
+
+	SECTION("for the core list case") { CHECK_THROWS_WITH(detail::thread_pinning::parse_validate_env("1,3#4"sv), Catch::Matchers::StartsWith(error_string)); }
+
+	SECTION("for random strings") { CHECK_THROWS_WITH(detail::thread_pinning::parse_validate_env("foo"sv), Catch::Matchers::StartsWith(error_string)); }
+}
+
+TEST_CASE("a warning is emitted if insufficient cores are available", "[affinity]") {
+	skip_unsupported_configurations();
+	test_utils::allow_max_log_level(detail::log_level::warn);
+	raii_affinity_masking mask({0, 1, 2});
+
+	SECTION("if pinning enabled") {
+		detail::thread_pinning::thread_pinner pinner({.enabled = true, .num_devices = 3, .num_legacy_processes = 1});
+		CHECK(test_utils::log_contains_substring(detail::log_level::warn, "Ran out of available cores for thread pinning"));
+	}
+	SECTION("if pinning disabled") {
+		detail::thread_pinning::thread_pinner pinner({.enabled = false, .num_devices = 3, .num_legacy_processes = 1});
+		CHECK(test_utils::log_contains_substring(detail::log_level::warn,
+		    fmt::format("Celerity has detected that only {} logical cores are available to this process. It is recommended to assign at least {}", 3, 6)));
+	}
+}
+
+TEST_CASE("do not plan for device submission threads if they are unused", "[affinity]") {
+	skip_unsupported_configurations();
+	raii_affinity_masking mask({1, 2, 3, 4});
+	const detail::thread_pinning::runtime_configuration cfg = {.enabled = true, .num_devices = 10, .use_backend_device_submission_threads = false};
+	detail::thread_pinning::thread_pinner pinner(cfg);
+	// no additional check, a warning will make the test fail if we do not handle this case correctly
```

Comment:
```suggestion
	SUCCEED(); // no additional check, a warning will make the test fail if we do not handle this case correctly
```

Reply:
Added. Out of curiosity, what does this change?